src/app.py

```diff
@@ -95,90 +103,94 @@ def query() -> Any:
+    )
 
 
-@app.route("/api/job/<hash>")
-def job_info(hash: str) -> Response:
-    return jsonify(redis.hgetall("job:" + hash))
+@app.get("/api/job/{job_id}", response_model=JobSchema)
+def job_info(job_id: str) -> JobSchema:
+    return get_job(f"job:{job_id}")
 
 
-@app.route("/api/job/<job_id>", methods=["DELETE"])
-def job_cancel(job_id: str) -> Response:
+@app.delete("/api/job/{job_id}", response_model=StatusSchema)
+def job_cancel(job_id: str) -> StatusSchema:
     redis.hmset("job:" + job_id, {"status": "cancelled"})
-
-    return jsonify({"status": "ok"})
+    return StatusSchema(status="ok")
 
 
-@app.route("/api/user/settings")
-def user_settings() -> Response:
-    return jsonify({"can_register": True, "plugin_name": "Redis"})
+@app.get("/api/user/settings", response_model=UserSettingsSchema)
+def user_settings() -> UserSettingsSchema:
+    return UserSettingsSchema(can_register=True, plugin_name="Redis")
 
 
-@app.route("/api/user/register", methods=["POST"])
-def user_register() -> Response:
-    if request.get_json()["username"].startswith("a"):
-        return jsonify({"status": "ok"})
-    else:
-        return jsonify({"error": "This user already exists"})
+@app.post("/api/user/register", response_model=StatusSchema)
+def user_register(auth: UserAuthSchema = Body(...)) -> StatusSchema:
+    if auth.username.startswith("a"):
+        return StatusSchema(status="ok")
+    raise HTTPException(status_code=400, detail="This user already exists")
 
 
-@app.route("/api/user/login", methods=["POST"])
-def user_login() -> Response:
-    if request.get_json()["username"].startswith("a"):
-        return jsonify({"status": "ok"})
-    else:
-        return jsonify({"error": "Wrong password"})
+@app.post("/api/user/login", response_model=StatusSchema)
+def user_login(auth: UserAuthSchema = Body(...)) -> StatusSchema:
+    if auth.username.startswith("a"):
+        return StatusSchema(status="ok")
+    raise HTTPException(status_code=400, detail="Wrong password")
 
 
-@app.route("/api/user/<name>/info")
-def user_info(name) -> Response:
-    return jsonify({"id": 1, "name": name})
+@app.get("/api/user/{name}/info", response_model=UserInfoSchema)
+def user_info(name: str) -> UserInfoSchema:
+    return UserInfoSchema(id=1, name=name)
 
 
-@app.route("/api/user/<name>/jobs")
-def user_jobs(name) -> Response:
+@app.get("/api/user/{name}/jobs", response_model=List[JobSchema])
+def user_jobs(name: str) -> List[JobSchema]:
     return job_statuses()
 
 
-@app.route("/api/job")
-def job_statuses() -> Response:
-    jobs = redis.keys("job:*")
-    jobs = sorted(
-        [dict({"id": job[4:]}, **redis.hgetall(job)) for job in jobs],
-        key=lambda o: o.get("submitted"),
-        reverse=True,
-    )
-
-    return jsonify({"jobs": jobs})
+@app.get("/api/job", response_model=JobsSchema)
+def job_statuses() -> JobsSchema:
+    jobs = [get_job(j) for j in redis.keys("job:*")]
+    return JobsSchema(jobs=sorted(jobs, key=lambda j: j.submitted))
```

Comment:
As above, return plain list

Reply:
I'm not too familiar with pydantic's schemas, but JobsSchema is a list of complex objects. How should I change this to better fit into fastapi/pydantic?

Reply:
You don't need to return a pydantic schema. You can just return a `List[JobSchema]` and it will work :)